### PR TITLE
Fix stack overflow in recursive Buffer tables.

### LINKED ISSUE #40637: SegFault with "recursive" buffer table
ClickHouse crushed on request to system.tables if has an incorrect "self-flushing" buffer table.

```
xxxx :) CREATE TABLE test (key UInt32) Engine=Buffer(default, test, 16, 10, 100, 10000, 1000000, 10000000, 100000000);


CREATE TABLE test
(
    `key` UInt32
)
ENGINE = Buffer(default, test, 16, 10, 100, 10000, 1000000, 10000000, 100000000)

Query id: a90ea965-ac7c-420f-a947-f900b6b0d014

Ok.

0 rows in set. Elapsed: 0.005 sec. 

xxxx :) select * from system.tables

SELECT *
FROM system.tables

Query id: bbb7d7eb-b782-4527-9803-ddc4f8703d5c

[xxxx] 2022.08.25 21:44:05.507879 [ 841325 ] <Fatal> BaseDaemon: ########################################
[xxxx] 2022.08.25 21:44:05.508036 [ 841325 ] <Fatal> BaseDaemon: (version 22.8.2.11 (official build), build id: F509FE73D83F8136) (from thread 841253) (query_id: bbb7d7eb-b782-4527-9803-ddc4f8703d5c) (query: select * from system.tables) Received signal Segmentation fault (11)
[xxxx] 2022.08.25 21:44:05.508105 [ 841325 ] <Fatal> BaseDaemon: Address: 0x7f06c12f7ff8 Access: write. Attempted access has violated the permissions assigned to the memory area.
[xxxx] 2022.08.25 21:44:05.508141 [ 841325 ] <Fatal> BaseDaemon: Stack trace: 0xa480065
[xxxx] 2022.08.25 21:44:05.508200 [ 841325 ] <Fatal> BaseDaemon: 2. ? @ 0xa480065 in /usr/bin/clickhouse
[xxxx] 2022.08.25 21:44:05.697982 [ 841325 ] <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: 26BBB329434F6C45E2D0312BCEE49FC1)
↘ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.)                                                                                                                                                   Exception on client:
Code: 32. DB::Exception: Attempt to read after eof: while receiving packet from localhost:9000. (ATTEMPT_TO_READ_AFTER_EOF)

Connecting to localhost:9000 as user default.
Code: 210. DB::NetException: Connection reset by peer, while writing to socket ([::1]:9000). (NETWORK_ERROR)

```

Found on version 22.3.9.19, reproduced also on 22.8.2.11

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix stack overflow in recursive `Buffer` tables. This closes #40637

### CODE REVIEW

File: src/Storages/StorageBuffer.cpp [hunk 1]
```diff
@@ -1010,14 +1012,12 @@ void StorageBuffer::checkAlterIsPossible(const AlterCommands & commands, Context
 std::optional<UInt64> StorageBuffer::totalRows(const Settings & settings) const
 {
     std::optional<UInt64> underlying_rows;
-    auto underlying = DatabaseCatalog::instance().tryGetTable(destination_id, getContext());
+    auto destination = DatabaseCatalog::instance().tryGetTable(destination_id, getContext());
 
-    if (underlying)
-        underlying_rows = underlying->totalRows(settings);
-    if (!underlying_rows)
-        return underlying_rows;
```
Comment (alexey-milovidov) [inline src/Storages/StorageBuffer.cpp:1018]: Here was another bug, CC @azat.

File: src/Storages/StorageBuffer.cpp [hunk 2]
```diff
@@ -1010,14 +1012,12 @@ void StorageBuffer::checkAlterIsPossible(const AlterCommands & commands, Context
 std::optional<UInt64> StorageBuffer::totalRows(const Settings & settings) const
 {
     std::optional<UInt64> underlying_rows;
-    auto underlying = DatabaseCatalog::instance().tryGetTable(destination_id, getContext());
+    auto destination = DatabaseCatalog::instance().tryGetTable(destination_id, getContext());
```
Comment (azat) [inline src/Storages/StorageBuffer.cpp:1015]: Maybe add a helper to get destination, like in #40640?

File: src/Storages/StorageBuffer.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/StorageBuffer.cpp:1018]: Indeed, nice catch. (Thread can be resolved).

File: tests/queries/0_stateless/02391_recursive_buffer.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test (key UInt32) Engine = Buffer(currentDatabase(), test, 16, 10, 100, 10000, 1000000, 10000000, 100000000);
+SELECT * FROM test; -- { serverError 269 }
+SELECT * FROM system.tables WHERE table = 'test' AND database = currentDatabase() FORMAT Null; -- { serverError 269 }
```
Comment (azat) [inline tests/queries/0_stateless/02391_recursive_buffer.sql:4]: Note, that concurrent run this test with `select * from system.tables` queries (in some other tests) may produce `INFINITE_LOOP` errors there, but maybe this test should be run in parallel, and instead tests that do `select * from system.tables` should not.

**Review by azat — APPROVED** [review]

**Comment by azat** [comment]
>[Stateless tests (ubsan) — Tests are not finished, fail: 1, passed: 63, skipped: 2](https://s3.amazonaws.com/clickhouse-test-reports/40643/f63fa3b04bc3321ebfbea685c5da0f946ae3e420/stateless_tests__ubsan_.html)

Simple queries takes too much time, I'm pretty sure that there are some hung tasks reported by the kernel, maybe dmesg should be captured for all tests (just in case).

>[Stress test (debug) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt)](https://s3.amazonaws.com/clickhouse-test-reports/40643/f63fa3b04bc3321ebfbea685c5da0f946ae3e420/stress_test__debug_.html)

Will take a look at OOM

>[Stress test (thread) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/40643/f63fa3b04bc3321ebfbea685c5da0f946ae3e420/stress_test__thread_.html)

```
==================
WARNING: ThreadSanitizer: data race (pid=666)
  Read of size 8 at 0x7b44023c74e0 by thread T2702 (mutexes: write M0):
    #0 <null> <null> (clickhouse+0x19f9e75b) (BuildId: 9ba41c32de4e62863b82c2387fcabcd793540e90)
```

Bad report, #40655 should help

There is a data-race between reading in `system.filesystem_cache`, and changing the segments for `FileSegment::range()`.
But the fix will conflict with #40586, so I will wait then (cc @kssenii).

```
  Read of size 8 at 0x7b44023c74e0 by thread T2702 (mutexes: write M0):
azat@s1:~/ch/tmp/40643$ llvm-symbolizer --exe ./clickhouse $(awk -F '[(+)]' '{print $3}' tsan-read.log) | head -n15
DB::FileSegment::Range::size() const
./build_docker/../src/Common/FileSegment.h:98:38
DB::FileSegment::getSnapshot(std::__1::shared_ptr<DB::FileSegment> const&, std::__1::lock_guard<std::__1::mutex>&)
./build_docker/../src/Common/FileSegment.cpp:667:31

DB::FileCache::getSnapshot() const
./build_docker/../src/Common/FileCache.cpp:1067:37

DB::StorageSystemFilesystemCache::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const
./build_docker/../src/Storages/System/StorageSystemFilesystemCache.cpp:43:37

DB::IStorageSystemOneBlock<DB::StorageSystemFilesystemCache>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned int)
./build_docker/../src/Storages/System/IStorageSystemOneBlock.h:54:9

DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned int)

  Previous write of size 8 at 0x7b44023c74e0 by thread T1807:

azat@s1:~/ch/tmp/40643$ llvm-symbolizer --exe ./clickhouse $(awk -F '[(+)]' '{print $3}' tsan-write.log) | head -n15
DB::FileSegmentRangeWriter::completeFileSegment(DB::FileSegment&)
./build_docker/../src/Common/FileSegment.cpp:827:36

DB::FileSegmentRangeWriter::finalize()
./build_docker/../src/Common/FileSegment.cpp:917:5

DB::CachedOnDiskWriteBufferFromFile::finalizeImpl()
./build_docker/../src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp:175:23

DB::WriteBuffer::finalize()
./build_docker/../src/IO/WriteBuffer.h:125:13

DB::MergedBlockOutputStream::Finalizer::Impl::finish()
./build_docker/../src/Storages/MergeTree/MergedBlockOutputStream.cpp:102:15
DB::MergedBlockOutputStream::Finalizer::finish()
```

**Comment by kssenii** [comment]
> There is a data-race between reading in system.filesystem_cache, and changing the segments for FileSegment::range().
But the fix will conflict with https://github.com/ClickHouse/ClickHouse/pull/40586, so I will wait then (cc @kssenii).

yeah, I fixed it in a different pr https://github.com/ClickHouse/ClickHouse/pull/40608.

**Comment by azat** [comment]
>[Stress test (undefined) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/40643/6940bdab3bb50ddb9db4362bd76c7b38439b822d/stress_test__undefined_.html)

- #40705

**Pull request merged**